smp_phy_control.8
.TH SMP_PHY_CONTROL "8" "July 2007" "smp_utils\-0.93" SMP_UTILS
.SH NAME
smp_phy_control \- invoke PHY CONTROL SMP function
.SH SYNOPSIS
.B smp_phy_control
[\fI\-\-attached=ADN\fR] [\fI\-\-expected=EX\fR] [\fI\-\-help\fR]
[\fI\-\-hex\fR] [\fI\-\-interface=PARAMS\fR] [\fI\-\-max=MA\fR]
[\fI\-\-min=MI\fR] [\fI\-\-op=OP\fR] [\fI\-\-phy=ID\fR] [\fI\-\-pptv=TI\fR]
[\fI\-\-raw\fR] [\fI\-\-sa=SAS_ADDR\fR] [\fI\-\-verbose\fR]
[\fI\-\-version\fR] \fISMP_DEVICE[,N]\fR
.SH DESCRIPTION
.\" Add any additional description here
.PP
Sends a SAS Management Protocol (SMP) PHY CONTROL request function to a SMP
target. The SMP target is identified by the \fISMP_DEVICE\fR and the
\fISAS_ADDR\fR. Depending on the interface, the \fISAS_ADDR\fR may be deduced
from the \fISMP_DEVICE\fR. With one interface there is one \fISMP_DEVICE\fR
per machine so the \fISMP_DEVICE,N\fR syntax is needed to differentiate
between HBAs if there are multiple present.
.PP
The PHY CONTROL function is used to change the state of a phy within a SMP
target. SMP targets are typically SAS expanders which have multiple phys.
Certain operation values (e.g. 'lr' (link reset) and 'hr' (hard reset))
change the state of the attached phy. Sending such operation values to the
phy in the SMP target that is attached to the originator (i.e. the SMP
initiator) may lead to a bad response.
.PP
Invoking this utility with no arguments (other than \fISMP_DEVICE\fR which
might be in an environment variable and \fI\-\-sa=SAS_ADDR\fR which might
be in an environment variable or not needed) is harmless. In other words a
phy's state is only changed when either \fI\-\-max=MA\fR, \fI\-\-min=MI\fR,
\fI\-\-op=OP\fR or \fI\-\-pptv=TI\fR is given with a non default value.
.SH OPTIONS
Mandatory arguments to long options are mandatory for short options as well.
.TP
\fB\-a\fR, \fB\-\-attached\fR=\fIADN\fR
specifies the attached device name (\fIADN\fR). The default value is 0 .
The \fIADN\fR is in decimal but is likely to be a SAS address which is
typically shown in hexadecimal. To specify a number in hexadecimal either
prefix it with '0x' or put a trailing 'h' on it.
.TP
\fB\-E\fR, \fB\-\-expected\fR=\fIEX\fR
set the 'expected expander change count' field in the SMP request.
The value \fIEX\fR is from 0 to 65535 inclusive with 0 being the default
value. When \fIEX\fR is greater than zero then if the value doesn't match
the expander change count of the SMP target (i.e. the expander) when
the request arrives then the target ignores the request and sets a
function result of "invalid expander change count" in the response.
.TP
\fB\-h\fR, \fB\-\-help\fR
output the usage message then exit.
.TP
\fB\-H\fR, \fB\-\-hex\fR
output the response in hexadecimal.
.TP
\fB\-I\fR, \fB\-\-interface\fR=\fIPARAMS\fR
interface specific parameters. In this case "interface" refers to the
path through the operating system to the SMP initiator. See the smp_utils
man page for more information.
.TP
\fB\-M\fR, \fB\-\-max\fR=\fIMA\fR
permits the programmed maximum physical link rate to be changed on the
gven phy. Permitted values are: 0 \-> no change, 8 \-> 1.5 Gbps, 9 \-> 3
Gbps, 10 \-> 6 Gbps. Default value is 0.
.TP
\fB\-m\fR, \fB\-\-min\fR=\fIMI\fR
permits the programmed minimum physical link rate to be changed on the
given phy. Permitted values are: 0 \-> no change, 8 \-> 1.5 Gbps, 9 \-> 3
Gbps, 10 \-> 6 Gbps. Default value is 0.
.TP
\fB\-o\fR, \fB\-\-op\fR=\fIOP\fR
specifies the operation to be performed on the given phy. The \fIOP\fR
argument can be either numeric or a string. If a number is given, it is put
into the 'phy operation' field of the request. Allowable strings are
abbreviations of which only the first two characters need to match. The
supported strings are: 'nop' (no operation), 'lr' (link reset), 'hr' (hard
reset), 'dis' (disable phy), 'cel' (clear error log), 'ca' (clear
affiliation), 'tspss' (transmit SATA port selection signal), 'citnl' (clear
STP I_T nexus loss (bit)), and 'sadn' (set attached device name).  The
default value is 0 (no operation).
.TP
\fB\-p\fR, \fB\-\-phy\fR=\fIID\fR
phy identifier. \fIID\fR is a value between 0 and 127. Default is 0.
.TP
\fB\-P\fR, \fB\-\-pptv\fR=\fITI\fR
partial pathway timeout value. The units are microseconds and the permitted
values are between 0 and 15 inclusive. 7 microseconds is recommended by
sas2r07.
.TP
\fB\-r\fR, \fB\-\-raw\fR
send the response to stdout in binary. All error messages are sent to stderr.
.TP
\fB\-s\fR, \fB\-\-sa\fR=\fISAS_ADDR\fR
specifies the SAS address of the SMP target device. Typically this is an
expander. This option may not be needed if the \fISMP_DEVICE\fR has the
target's SAS address within it. The \fISAS_ADDR\fR is in decimal but most SAS
addresses are shown in hexadecimal. To give a number in hexadecimal either
prefix it with '0x' or put a trailing 'h' on it.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
increase the verbosity of the output. Can be used multiple times
.TP
\fB\-V\fR, \fB\-\-version\fR
print the version string and then exit.
.SH AUTHORS
Written by Douglas Gilbert.
.SH "REPORTING BUGS"
Report bugs to <dgilbert at interlog dot com>.
.SH COPYRIGHT
Copyright \(co 2006\-2007 Douglas Gilbert
.br
This software is distributed under a FreeBSD license. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.B smp_utils, smp_discover(smp_utils)


smp_phy_control.c
/*
 * Copyright (c) 2006-2007 Douglas Gilbert.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */

#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <ctype.h>
#include <getopt.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "smp_lib.h"

/* This is a Serial Attached SCSI (SAS) management protocol (SMP) utility
 * program.
 *
 * This utility issues a PHY CONTROL function and outputs its response.
 */

static char * version_str = "1.08 20070707";


static struct option long_options[] = {
        {"attached", 1, 0, 'a'},
        {"expected", 1, 0, 'E'},
        {"help", 0, 0, 'h'},
        {"hex", 0, 0, 'H'},
        {"interface", 1, 0, 'I'},
        {"min", 1, 0, 'm'},
        {"max", 1, 0, 'M'},
        {"op", 1, 0, 'o'},
        {"phy", 1, 0, 'p'},
        {"pptv", 1, 0, 'P'},
        {"raw", 0, 0, 'r'},
        {"sa", 1, 0, 's'},
        {"verbose", 0, 0, 'v'},
        {"version", 0, 0, 'V'},
        {0, 0, 0, 0},
};

static void usage()
{
    fprintf(stderr, "Usage: "
          "smp_phy_control [--attached=ADN] [--expected=EX] [--help] "
          "[--hex]\n"
          "                       [--interface=PARAMS] [--max=MA] "
          "[--min=MI] [--op=OP]\n"
          "                       [--phy=ID] [pptv=TI] [--raw] "
          "[--sa=SAS_ADDR]\n"
          "                       [--verbose] [--version] "
          "SMP_DEVICE[,N]\n"
          "  where:\n"
          "    --attached=ADN|-a ADN    attached device name\n"
          "    --expected=EX|-E EX    set expected expander change count "
          "to EX\n"
          "    --help|-h            print out usage message\n"
          "    --hex|-H             print response in hexadecimal\n"
          "    --interface=PARAMS|-I PARAMS    specify or override "
          "interface\n"
          "    --max=MA|-M MA       programmable maximum physical link "
          "speed\n"
          "                         (8->1.5 Gbps, 9->3 Gbps, "
          "10->6 Gbps)\n"
          "    --min=MI|-m MI       programmable minimum physical link "
          "speed\n"
          "    --op=OP|-o OP        OP (operation) is a number or "
          "abbreviation (from:\n"
          "                         nop, lr, hr, dis, cel, ca, tspss, "
          "citnl, sadn)\n"
          "    --phy=ID|-p ID       phy identifier (def: 0)\n"
          "    --pptv=TI|-P TI      partial pathway timeout value "
          "(microseconds)\n"
          "                         (if given sets UPPTV bit)\n"
          "    --raw|-r             output response in binary\n"
          "    --sa=SAS_ADDR|-s SAS_ADDR    SAS address of SMP "
          "target (use leading '0x'\n"
          "                         or trailing 'h'). Depending on "
          "the interface, may\n"
          "                         not be needed\n"
          "    --verbose|-v         increase verbosity\n"
          "    --version|-V         print version string and exit\n\n"
          "Performs a SMP PHY CONTROL function\n"
          );

}

static void dStrRaw(const char* str, int len)
{
    int k;

    for (k = 0 ; k < len; ++k)
        printf("%c", str[k]);
}

static struct smp_val_name op_abbrev[] = {
    {0, "nop"},
    {1, "lr"},          /* link reset */
    {2, "hr"},          /* hard reset */
    {3, "dis"},         /* disable phy */
    {5, "cel"},         /* clear error log */
    {6, "ca"},          /* clear affiliation */
    {7, "tspss"},       /* transmit SATA port selection signal */
    {8, "citnl"},       /* clear STP I_T nexus loss */
    {9, "sadn"},        /* set attached device name */
    {-1, NULL},
};

static void list_op_abbrevs()
{
    struct smp_val_name * vnp;

    fprintf(stderr, "  Valid operation abbreviations are:\n");
    for (vnp = op_abbrev; vnp->name; ++vnp)
        fprintf(stderr, "    %s\n", vnp->name);
}


int main(int argc, char * argv[])
{
    int res, c, k, j, len;
    int expected_cc = 0;
    int do_hex = 0;
    int do_min = 0;
    int do_max = 0;
    int op_val = 0;
    int pptv = -1;
    int phy_id = 0;
    int do_raw = 0;
    int verbose = 0;
    long long sa_ll;
    unsigned long long sa = 0;
    unsigned long long adn = 0;
    char i_params[256];
    char device_name[512];
    char b[256];
    unsigned char smp_req[] = {SMP_FRAME_TYPE_REQ, SMP_FN_PHY_CONTROL, 0, 9,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, };
    unsigned char smp_resp[128];
    struct smp_req_resp smp_rr;
    struct smp_target_obj tobj;
    struct smp_val_name * vnp;
    int subvalue = 0;
    char * cp;
    int ret = 0;

    memset(device_name, 0, sizeof device_name);
    while (1) {
        int option_index = 0;

        c = getopt_long(argc, argv, "a:E:hHI:m:M:o:p:P:rs:vV", long_options,
                        &option_index);
        if (c == -1)
            break;

        switch (c) {
        case 'a':
           sa_ll = smp_get_llnum(optarg);
           if (-1LL == sa_ll) {
                fprintf(stderr, "bad argument to '--attached'\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
            adn = (unsigned long long)sa_ll;
            break;
        case 'E':
            expected_cc = smp_get_num(optarg);
            if ((expected_cc < 0) || (expected_cc > 65535)) {
                fprintf(stderr, "bad argument to '--expected'\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
            break;
        case 'h':
        case '?':
            usage();
            return 0;
        case 'H':
            ++do_hex;
            break;
        case 'I':
            strncpy(i_params, optarg, sizeof(i_params));
            i_params[sizeof(i_params) - 1] = '\0';
            break;
        case 'm':
            do_min = smp_get_num(optarg);
            switch (do_min) {
            case 0:
            case 8:
            case 9:
            case 10:
                break;
            default:
                fprintf(stderr, "bad argument to '--min', want 0, 8, 9 "
                        "or 10\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
            break;
        case 'M':
            do_max = smp_get_num(optarg);
            switch (do_max) {
            case 0:
            case 8:
            case 9:
            case 10:
                break;
            default:
                fprintf(stderr, "bad argument to '--max', want 0, 8, 9 "
                        "or 10\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
            break;
        case 'o':
            if (isalpha(optarg[0])) {
                for (vnp = op_abbrev; vnp->name; ++vnp) {
                    if (0 == strncmp(optarg, vnp->name, 2))
                        break;
                }
                if (vnp->name)
                    op_val = vnp->value;
                else {
                    fprintf(stderr, "bad argument to '--op'\n");
                    list_op_abbrevs();
                    return SMP_LIB_SYNTAX_ERROR;
                }
            } else {
                op_val = smp_get_num(optarg);
                if ((op_val < 0) || (op_val > 255)) {
                    fprintf(stderr, "bad numeric argument to '--op'\n");
                    return SMP_LIB_SYNTAX_ERROR;
                }
            }
            break;
        case 'p':
           phy_id = smp_get_num(optarg);
           if ((phy_id < 0) || (phy_id > 127)) {
                fprintf(stderr, "bad argument to '--phy'\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
            break;
        case 'P':
           pptv = smp_get_num(optarg);
           if ((pptv < 0) || (pptv > 15)) {
                fprintf(stderr, "bad argument to '--pptv', want value"
                        "from 0 to 15 inclusive\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
            break;
        case 'r':
            ++do_raw;
            break;
        case 's':
           sa_ll = smp_get_llnum(optarg);
           if (-1LL == sa_ll) {
                fprintf(stderr, "bad argument to '--sa'\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
            sa = (unsigned long long)sa_ll;
            break;
        case 'v':
            ++verbose;
            break;
        case 'V':
            fprintf(stderr, "version: %s\n", version_str);
            return 0;
        default:
            fprintf(stderr, "unrecognised switch code 0x%x ??\n", c);
            usage();
            return SMP_LIB_SYNTAX_ERROR;
        }
    }
    if (optind < argc) {
        if ('\0' == device_name[0]) {
            strncpy(device_name, argv[optind], sizeof(device_name) - 1);
            device_name[sizeof(device_name) - 1] = '\0';
            ++optind;
        }
        if (optind < argc) {
            for (; optind < argc; ++optind)
                fprintf(stderr, "Unexpected extra argument: %s\n",
                        argv[optind]);
            usage();
            return SMP_LIB_SYNTAX_ERROR;
        }
    }
    if (0 == device_name[0]) {
        cp = getenv("SMP_UTILS_DEVICE");
        if (cp)
            strncpy(device_name, cp, sizeof(device_name) - 1);
        else {
            fprintf(stderr, "missing device name!\n    [Could use "
                    "environment variable SMP_UTILS_DEVICE instead]\n");
            usage();
            return SMP_LIB_SYNTAX_ERROR;
        }
    }
    if ((cp = strchr(device_name, ','))) {
        *cp = '\0';
        if (1 != sscanf(cp + 1, "%d", &subvalue)) {
            fprintf(stderr, "expected number after comma in SMP_DEVICE "
                    "name\n");
            return SMP_LIB_SYNTAX_ERROR;
        }
    }
    if (0 == sa) {
        cp = getenv("SMP_UTILS_SAS_ADDR");
        if (cp) {
           sa_ll = smp_get_llnum(cp);
           if (-1LL == sa_ll) {
                fprintf(stderr, "bad value in environment variable "
                        "SMP_UTILS_SAS_ADDR\n");
                fprintf(stderr, "    use 0\n");
                sa_ll = 0;
            }
            sa = (unsigned long long)sa_ll;
        }
    }
    if (sa > 0) {
        if (! smp_is_naa5(sa)) {
            fprintf(stderr, "SAS (target) address not in naa-5 format "
                    "(may need leading '0x')\n");
            if ('\0' == i_params[0]) {
                fprintf(stderr, "    use '--interface=' to override\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
        }
    }

    res = smp_initiator_open(device_name, subvalue, i_params, sa,
                             &tobj, verbose);
    if (res < 0)
        return SMP_LIB_FILE_ERROR;

    smp_req[4] = (expected_cc >> 8) & 0xff;
    smp_req[5] = expected_cc & 0xff;
    smp_req[9] = phy_id;
    smp_req[10] = op_val;
    if (pptv >= 0) {
        smp_req[11] |= 1;
        smp_req[36] |= (pptv & 0xf);
    }
    if (adn) {
        for (j = 0; j < 8; ++j, (adn >>= 8))
            smp_req[24 + 7 - j] = (adn & 0xff);
    }
    smp_req[32] |= (do_min << 4);
    smp_req[33] |= (do_max << 4);
    if (verbose) {
        fprintf(stderr, "    Phy control request: ");
        for (k = 0; k < (int)sizeof(smp_req); ++k)
            fprintf(stderr, "%02x ", smp_req[k]);
        fprintf(stderr, "\n");
    }

    memset(&smp_rr, 0, sizeof(smp_rr));
    smp_rr.request_len = sizeof(smp_req);
    smp_rr.request = smp_req;
    smp_rr.max_response_len = sizeof(smp_resp);
    smp_rr.response = smp_resp;
    res = smp_send_req(&tobj, &smp_rr, verbose);

    if (res) {
        fprintf(stderr, "smp_send_req failed, res=%d\n", res);
        if (0 == verbose)
            fprintf(stderr, "    try adding '-v' option for more debug\n");
        ret = -1;
        goto err_out;
    }
    if (smp_rr.transport_err) {
        fprintf(stderr, "smp_send_req transport_error=%d\n",
                smp_rr.transport_err);
        ret = -1;
        goto err_out;
    }
    if ((smp_rr.act_response_len >= 0) && (smp_rr.act_response_len < 4)) {
        fprintf(stderr, "response too short, len=%d\n", smp_rr.act_response_len);
        ret = SMP_LIB_CAT_MALFORMED;
        goto err_out;
    }
    len = smp_resp[3];
    if (0 == len) {
        len = smp_get_func_def_resp_len(smp_resp[1]);
        if (len < 0) {
            len = 0;
            if (verbose > 0)
                fprintf(stderr, "unable to determine response length\n");
        }
    }
    len = 4 + (len * 4);        /* length in bytes, excluding 4 byte CRC */
    if (do_hex || do_raw) {
        if (do_hex)
            dStrHex((const char *)smp_resp, len, 1);
        else
            dStrRaw((const char *)smp_resp, len);
        if (SMP_FRAME_TYPE_RESP != smp_resp[0])
            ret = SMP_LIB_CAT_MALFORMED;
        if (smp_resp[1] != smp_req[1])
            ret = SMP_LIB_CAT_MALFORMED;
        if (smp_resp[2])
            ret = smp_resp[2];
        goto err_out;
    }
    if (SMP_FRAME_TYPE_RESP != smp_resp[0]) {
        fprintf(stderr, "expected SMP frame response type, got=0x%x\n", smp_resp[0]);
        ret = SMP_LIB_CAT_MALFORMED;
        goto err_out;
    }
    if (smp_resp[1] != smp_req[1]) {
        fprintf(stderr, "Expected function code=0x%x, got=0x%x\n", smp_req[1], smp_resp[1]);
        ret = SMP_LIB_CAT_MALFORMED;
        goto err_out;

    }
    if (smp_resp[2]) {
        cp = smp_get_func_res_str(smp_resp[2], sizeof(b), b);
        fprintf(stderr, "Phy control result: %s\n", cp);
        ret = smp_resp[2];
        goto err_out;
    }

err_out:
    res = smp_initiator_close(&tobj);
    if (res < 0) {
        fprintf(stderr, "close error: %s\n", safe_strerror(errno));
        if (0 == ret)
            return SMP_LIB_FILE_ERROR;
    }
    return (ret >= 0) ? ret : SMP_LIB_CAT_OTHER;
}
